File: zircon/kernel/object/sampler_dispatcher.cc
// Copyright 2026 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

#include <lib/thread_sampler/thread_sampler.h>

#include <object/sampler_dispatcher.h>

void SamplerDispatcher::on_zero_handles() {
  if (zx::result res = sampler::gThreadSampler.Destroy(); res.is_error()) {
    dprintf(ALWAYS, "Failed to cleanly destroy sampler: %d\n", res.status_value());
  }
}

zx::result<KernelHandle<SamplerDispatcher>> SamplerDispatcher::Create(
    const zx_sampler_config_t& config) {
  // Set up the global sampler if it hasn't been set up yet.
  zx::result res = sampler::gThreadSampler.SetUp(config);
  if (res.is_error()) {
    return res.take_error();
  }

  fbl::AllocChecker ac;
  KernelHandle sampler(fbl::AdoptRef(new (&ac) SamplerDispatcher));
  if (!ac.check()) {
    return zx::error(ZX_ERR_NO_MEMORY);
  }

  return zx::ok(ktl::move(sampler));
}

zx::result<> SamplerDispatcher::Stop() { return sampler::gThreadSampler.Stop(); }

zx::result<> SamplerDispatcher::Start() { return sampler::gThreadSampler.Start(); }

zx::result<> SamplerDispatcher::SampleThread(zx_koid_t pid, zx_koid_t tid, GeneralRegsSource source,
                                             const void* gregs) {
  return sampler::gThreadSampler.SampleThread(pid, tid, source, gregs);
}

ktl::pair<zx_status_t, size_t> SamplerDispatcher::ReadUser(user_out_ptr<void> ptr, size_t len) {
  // We unfortunately run into some complexity here: while the buffer our samples in is created by
  // the kernel and is safe to read from, the user memory we are writing to could be pager-backed.
  // This means that when we attempt to write to it as part of the VmObjectPaged::ReadUser call, we
  // cannot be holding locks. So we need to obtain the lock to set up the copy, drop the lock, do
  // the copy, then grab the lock again to make sure everything went well.
  //
  // During the copy, we'd need to prevent:
  //   1) The sampler from writing new data
  //   2) The buffers being destroyed due to the read handle being zx_handle_close'd
  //   3) A new sampler from being created.
  //
  // We do this by:
  //    1) Setting our state to SamplingState::Reading which disallows starting a new session (and
  //       thus destroying the old one).
  //    2) If on_zero_handles is triggered while in `Reading` mode, we delay actually
  //       destroying the buffers and destroy them after the copy is completed instead.
  zx::result<sampler::ReadToken> token = sampler::gThreadSampler.PrepareRead();
  if (token.is_error()) {
    return {token.error_value(), 0};
  }

  auto [status, read] = sampler::gThreadSampler.ReadUser(*token, ptr, len);

  // We now need to ensure that the user side handle hasn't been dropped. If it has been, then
  // we need to clean it up.
  sampler::gThreadSampler.FinishRead(ktl::move(*token));
  return {status, read};
}

File: zircon/kernel/object/sampler_dispatcher.rs
// Copyright 2026 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

use crate::user_copy::UserOutPtr;
use core::mem::MaybeUninit;
use counters_rs::define_kcounter;
use fbl::Canary;
use ksync::{KMutex, RawCriticalMutex, guarded};
use pin_init::{PinInit, pin_data, pin_init, pinned_drop};
use zx_status::Status;
use zx_types::{
    ZX_OBJ_TYPE_SAMPLER, ZX_RIGHT_DUPLICATE, ZX_RIGHT_INSPECT, ZX_RIGHT_TRANSFER, zx_rights_t,
    zx_sampler_config_t,
};

use super::KernelHandle;
use super::sampler_dispatcher_ffi::cpp_sampler_dispatcher_create;

use object_constants_rs as object_constants;

/// Default rights assigned to a newly created SamplerDispatcher handle.
pub const DEFAULT_RIGHTS: zx_rights_t = ZX_RIGHT_TRANSFER | ZX_RIGHT_DUPLICATE | ZX_RIGHT_INSPECT;

zr::static_assert_size_and_align!(
    SamplerDispatcherState,
    object_constants::kSamplerDispatcherStateSize,
    object_constants::kSamplerDispatcherStateAlign,
);

define_kcounter!(DISPATCHER_SAMPLER_CREATE_COUNT, "dispatcher.sampler.create", Sum);
define_kcounter!(DISPATCHER_SAMPLER_DESTROY_COUNT, "dispatcher.sampler.destroy", Sum);

/// Internal state storage for `SamplerDispatcher`.
#[guarded]
#[pin_data(PinnedDrop)]
#[repr(C)]
pub struct SamplerDispatcherState {
    canary: Canary<{ fbl::magic(b"SAMP") }>,

    #[mutex]
    lock: KMutex<RawCriticalMutex>,
}

impl SamplerDispatcherState {
    /// Initializes the `SamplerDispatcherState`.
    pub fn init(
        _dispatcher: *const SamplerDispatcher,
    ) -> impl PinInit<Self, core::convert::Infallible> {
        DISPATCHER_SAMPLER_CREATE_COUNT.add(1);
        pin_init!(Self {
            canary: Canary::new(),
            lock <- KMutex::init(),
        })
    }
}

#[pinned_drop]
impl PinnedDrop for SamplerDispatcherState {
    fn drop(self: core::pin::Pin<&mut Self>) {
        DISPATCHER_SAMPLER_DESTROY_COUNT.add(1);
    }
}

crate::object::dispatcher::impl_dispatcher_facade_with_state!(
    pub struct SamplerDispatcher,
    SamplerDispatcherState,
    ZX_OBJ_TYPE_SAMPLER,
    object_constants::kSamplerDispatcherStateOffset
);

impl SamplerDispatcher {
    /// Returns the default rights for a SamplerDispatcher handle.
    pub fn default_rights() -> zx_rights_t {
        DEFAULT_RIGHTS
    }

    /// Creates a new `SamplerDispatcher` with the given configuration.
    pub fn create(
        config: &zx_sampler_config_t,
    ) -> Result<(KernelHandle<Self>, zx_rights_t), Status> {
        let rights = Self::default_rights();
        let mut handle_out = MaybeUninit::<KernelHandle<Self>>::uninit();
        // SAFETY: `config` is a valid reference and `handle_out` points to uninitialized handle
        // memory.
        let status = unsafe { cpp_sampler_dispatcher_create(config, &raw mut handle_out) };
        Status::ok(status)?;
        // SAFETY: cpp_sampler_dispatcher_create initialized handle_out.
        unsafe { Ok((handle_out.assume_init(), rights)) }
    }

    /// Starts sampling session for this sampler.
    pub fn start(&self) -> Result<(), Status> {
        // SAFETY: `self` is a valid SamplerDispatcher pointer.
        let status = unsafe {
            super::sampler_dispatcher_ffi::cpp_sampler_dispatcher_start(self as *const _)
        };
        Status::ok(status)
    }

    /// Stops sampling session for this sampler.
    pub fn stop(&self) -> Result<(), Status> {
        // SAFETY: `self` is a valid SamplerDispatcher pointer.
        let status =
            unsafe { super::sampler_dispatcher_ffi::cpp_sampler_dispatcher_stop(self as *const _) };
        Status::ok(status)
    }

    /// Read out the data contained in the sampler buffers into `ptr` return the number of bytes
    /// written. The Sampling state must be Stopped before calling this function.
    ///
    /// `len` _must_ be at least equal to the total size of the sampler buffers, which can be
    /// queried by passing a null `ptr`. In this case, no data will be written and the return value
    /// will be the required minimum size of the buffer to write to.
    pub fn read_user(&self, ptr: UserOutPtr<u8>, len: usize) -> (Result<(), Status>, usize) {
        let mut actual = 0;
        // SAFETY: `self` is valid, `ptr` is a valid UserOutPtr buffer of `len` bytes, and
        // `actual` is a local out pointer.
        let status = unsafe {
            super::sampler_dispatcher_ffi::cpp_sampler_dispatcher_read_user(
                self as *const _,
                ptr.as_ptr().cast(),
                len,
                &mut actual,
            )
        };
        (Status::ok(status), actual)
    }
}
